README.rst
.. image:: https://travis-ci.org/arskom/spyne.png?branch=master
        :target: http://travis-ci.org/arskom/spyne

**WARNING:** This is from spyne's development branch. This version is not released
yet! Latest stable release can be found in the ``2_10`` branch.

About
=====

Spyne aims to save the protocol implementers the hassle of implementing their
own remote procedure call api and the application programmers the hassle of
jumping through hoops just to expose their services using multiple protocols and
transports.

In other words, Spyne is a framework for building distributed
solutions that strictly follow the MVC pattern, where Model = `spyne.model`,
View = `spyne.protocol` and Controller = `user code`.

Spyne comes with the implementations of popular transport, protocol and
interface document standards along with a well-defined API that lets you
build on existing functionality.

Spyne currently supports the WSDL 1.1 interface description standard, along with
SOAP 1.1 and the so-called HttpRpc, XmlDocument, JsonDocument, YamlDocument,
MessagePackDocument and MessagePackRpc protocols which can be transported via Http
or ZeroMQ. The transports can be used in both a client or server setting.

The following are the primary sources of information about spyne:

* Spyne's home page (that also has the latest documentation) is: http://spyne.io/
* The latest documentation for Spyne can be found at: http://spyne.io/docs
* The official source code repository is at: https://github.com/arskom/spyne
* The official spyne discussion forum is at: http://mail.python.org/mailman/listinfo/soap
* You can download Spyne releases from `github <http://github.com/arskom/spyne/downloads>`_
  or `pypi <http://pypi.python.org/pypi/spyne>`_.

Spyne is a generalized version of a Soap library known as soaplib. The following
legacy versions of soaplib are also available in the source repository at github
as branches:

* Soaplib-0.8 branch: http://github.com/arskom/spyne/tree/soaplib-0_8
* Soaplib-1.0 branch: http://github.com/arskom/spyne/tree/soaplib-1_0
* Soaplib-2.0 was never released as a stable package, but the branch is still
  available: http://github.com/arskom/spyne/tree/soaplib-2_0

Requirements
============

Spyne is known to work on Python versions 2.6 and 2.7. We're also looking for
volunteers to test Python 3.x.

The only hard requirement is `pytz <http://pytz.sourceforge.net/>`_ which is
available via pypi.

Additionally the following software packages are needed for various subsystems
of Spyne:

* A Wsgi server of your choice is needed to wrap
  ``spyne.server.wsgi.WsgiApplication``
* `lxml>=2.3 <http://lxml.de>`_ is needed for any xml or html related protocol.
* `SQLAlchemy <http://sqlalchemy.org>`_ is needed for
  ``spyne.model.complex.TTableModel``.
* `pyzmq <https://github.com/zeromq/pyzmq>`_ is needed for
  ``spyne.client.zeromq.ZeroMQClient`` and
  ``spyne.server.zeromq.ZeroMQServer``.
* `Werkzeug <http://werkzeug.pocoo.org/>`_ is needed for using
  ``spyne.protocol.http.HttpRpc`` under a wsgi transport.
* `PyParsing<2.0 <http://pypi.python.org/pypi/pyparsing>`_ is needed for
  using ``HttpPattern``'s with ``spyne.protocol.http.HttpRpc``\. (PyParsing>=2.x
  is Python 3 only).
* `Twisted <http://twistedmatrix.com/>`_ is needed for
  ``spyne.server.twisted.TwistedWebResource`` and
  ``spyne.client.twisted.TwistedHttpClient``.
* `Django <http://djangoproject.com/>`_ (tested with 1.2 and up) is needed for
  :class:`spyne.server.django.DjangoApplication` and :class:`spyne.server.django.DjangoServer`.
* `Pyramid <http://pylonsproject.org/>`_ is needed for
  ``spyne.server.pyramid.PyramidApplication``.
* `msgpack-python <http://github.com/msgpack/msgpack-python/>`_ is needed for
  ``spyne.protocol.msgpack``.
* `PyYaml <https://bitbucket.org/xi/pyyaml>`_ is needed for
  ``spyne.protocol.yaml``.
* `simplejson <http://github.com/simplejson/simplejson>`_ is used when found
  for ``spyne.protocol.json``.

You are advised to add these as requirements to your own projects, as these are
only optional dependencies of Spyne, thus not handled in its setup script.

Installing
==========

You can get spyne via pypi: ::

    easy_install spyne

or you can clone from github: ::

    git clone git://github.com/arskom/spyne.git

or get the source distribution from one of the download sites and unpack it.

To install from source distribution, you should run its setup script as usual: ::

    python setup.py install [--user]

To run the tests use: ::

    pyhon setup.py test

The test script first installs every single library that Spyne integrates with,
in the current directory, so be ready to do some fiddling with your distro's
package manager in case you don't want this, or have a fully functional python
development environment ready so that packages like lxml can compile.

Finally if you want to make any changes to the Spyne code, just use ::

    python setup.py develop [--user]

so that you can painlessly test your patches.


Getting Support
===============

The main developers of Spyne lurk in the `official soap implementors
forum <http://mail.python.org/mailman/listinfo/soap/>`_ kindly operated
by python.org. That's mostly because Spyne is the continuation of soaplib,
but also because Soap is an important part of Spyne.

You can also use the 'spyne' tag to ask questions on
`Stack Overflow <http://stackoverflow.com>`_.


Contributing
============

If you wish to contribute to Spyne's development, create a personal fork
on GitHub.  When you are ready to push to the upstream repository,
submit a pull request to bring your work to the attention of the core
committers. They will respond to review your patch and act accordingly.

To save both parties time, make sure the existing tests pass. If you are
adding new functionality or fixing a bug, please have the accompanying test.
This will both help us increase test coverage and insure your use-case
is immune to feature code changes. You could also summarize in one or
two lines how your work will affect the life of Spyne users in the
CHANGELOG file.

Please follow the `PEP 8 <http://www.python.org/dev/peps/pep-0008/>`_
style guidelines for both source code and docstrings.

We could also use help with the docs, which are built from
`restructured text <http://docutils.sourceforge.net/rst.html>`_ using
`Sphinx <http://sphinx.pocoo.org>`_.

Regular contributors may be invited to join as a core Spyne committer on
GitHub. Even if this gives the core committers the power to commit directly
to the core repository, we highly value code reviews and expect every
significant change to be committed via pull requests.


A small note about submitting Pull Requests
-------------------------------------------

Github's pull-request feature is awesome, but there's a subtlety that's not
totally obvious for newcomers: If you continue working on the branch that you
used to submit a pull request, your commits will "pollute" the pull request
until it gets merged. This is not a bug, but a feature -- it gives you the
ability to address reviewers' concerns without creating pull requests over and
over again. So, if you intend to work on other parts of spyne after submitting
a pull request, please do move your work to its own branch and never submit a
pull request from your master branch. This will give you the freedom to
continue working on Spyne while waiting for your pull request to be reviewed.


CHANGELOG.rst

Changelog
=========

spyne-2.11.0
------------
* Add support for altering output protocol by setting ``ctx.out_protocol``.
* Add returning ctx.out_string support to null server (The ``ostr`` argument).
* Add support for XmlData modifier. It lets mapping the data in the xml body
  to an object field.
* Remove deprecated ``JsonObject`` identifier. Just do a gentle
  ``s/JsonObject/JsonDocument/g`` if you're still using it.
* SQLAlchemy: Implement storing arrays of simple types in a table.
* SQLAlchemy: Make it work with multiple foreign keys from one table to
  another.
* Implement an Xml Schema parser.
* Import all model markers to ``spyne.model``.
* Import ``@rpc``\, ``@srpc``\, ``ServiceBase`` and ``Application`` inside the
  ``spyne`` module.
* Implement JsonP protocol.
* Implement SpyneJsonRpc 1.0 protocol -- it supports request headers.

  Sample request:  ``{"ver":1, "body": {"div": [4,2]}, "head": {"id": 1234}}``
  Sample response: ``{"ver":1, "body": 2}``

  Sample request:  ``{"ver":1, "body": {"div": {"dividend":4,"divisor":0]}}``
  Sample response: ``{"ver":1, "fault": {"faultcode": "Server", "faultstring": "Internal Error"}}}``

* Steal and integrate the WebSocket tranport for Twisted.
* Support Django natively using :class:`spyne.server.django.SpyneView` and
  :class:`spyne.server.django.DjangoServer`.
* It's now possible to override the ``JsonEncoder`` class ``JsonDocument`` uses.
* Remove hard-coded utf-8 defaults from almost everywhere.
* Remove hard-coded pytz.utc defaults from everywhere. Use spyne.LOCAL_TZ to
  configure the default time zone.
* As a result of the above change, datetime objects deserialized by Spyne always
  carry time zones. Make sure you adjust your code to this change. It's possible
  to strip timezone information from outgoing types by passing ``timezone=False``
  to the DateTime customizer.
* Add default_factory to ModelBase customizer. It's a callable that produces
  default values on demand. Suitable to be used with e.g. lambdas that return
  mutable defaults.
* Add mapper between Django models and :class:`spyne.util.django.DjangoComplexModel`
  types.

spyne-2.10.9
------------
* Fix total_seconds quirk for Python 2.6.
* Turn off Xml features like entity resolution by default. This mitigates
  an information disclosure attack risk in services whose response contain
  some fragments or all of the request. Also prevents DoS attacks that make use
  of entity expansion. See https://bitbucket.org/tiran/defusedxml for more info.
* Drop Python 2.5 support (It wasn't working anyway).

spyne-2.10.8
------------
* Fix Unicode losing pattern on re-customization
* Fix Duration serialization, add a ton of test cases.
* Fix binary urlsafe_base64 encoding.
* Fix arbitrary exception serialization.
* Fix some doc errors.

spyne-2.10.7
------------
* Fix logic error in wsdl caching that prevented the url in Wsdl document from
  being customized.
* Fix dictdoc not playing well with functions with empty return values.

spyne-2.10.6
------------
* Fix exception serialization regression in DictDocument family.
* Fix xml utils (and its example).

spyne-2.10.5
------------
* Fix default value handling in ``HttpRpc``.
* Fix invalid document type raising ``InternalError`` in DictDocument family.
  It now raises ``ValidationError``.
* HttpRpc: Fix ``ByteArray`` deserialization.
* HttpRpc: Fix many corner cases with ``Array``\s.
* Fix Csv serializer.
* Fix Mandatory variants of ``Double`` and ``Float`` inheriting from decimal.

spyne-2.10.4
------------
* Fix handling of ``spyne.model.binary.File.Value`` with just path name.
* Fix decimal restrictions (some more).
* Make user code that doesn't return anything work with twisted server
  transport.

spyne-2.10.3
------------
* Add validation tests for HierDictDocument and fix seen issues.
* Add validation tests for FlatDictDocument and fix seen issues.
* Clarify Json and Http behavior in relevant docstrings.
* Fix Python2.6 generating max_occurs="inf" instead of "unbounded" sometimes.

spyne-2.10.2
------------
* Fix ByteArray support accross all protocols.
* Fix namespaces of customized simple types inside ``XmlAttribute`` not being
  imported.

spyne-2.10.1
------------
* Fix confusion in Decimal restriction assignment.
* Fix classmethod calls to ProtocolBase.
* Fix schema generation error in namespaced xml attribute case.

spyne-2.10.0
------------
* Returning twisted's Deferred from user code is now supported.
* You can now set Http response headers via ctx.out_header when
  out_protocol is HttpRpc. https://github.com/arskom/spyne/pull/201
* lxml is not a hard requirement anymore.
* XmlDocument and friends: cleanup_namespaces is now True by default.
* XmlDocument and friends: Added ``encoding`` and ``pretty_print`` flags that
  are directly passed to ``lxml.etree.tostring()``.
* XmlDocument and friends:'attribute_of' added to ModelBase to add attribute
  support for primitives. This is currently ignored by (and mostly irrelevant
  to) other protocols.
* XmlDocument and friends: Attribute serialization is working for arrays.
* Add support for exposing existing whose source code via the _args argument
  to the srpc decorator. See the existing_api example for usage examples.
* Add Streaming versions of Pyramid and Django bridge objects.
* Remove destructor from ``MethodContext``. Now transports need to call
  ``.close()`` explicitly to close object and fire relevant events.
* Application event 'method_context_constructed' was renamed to
  ``'method_context_created'``.
* Application event 'method_context_destroyed' was removed. The
  ``'method_context_closed'`` event can be used instead.
* SQLAlchemy integration now supports advanced features like specifying
  indexing methods.
* The object composition graph can now be cyclic.
* Integers were overhauled. Now boundary values of limited-size types are
  accessible via ``Attributes._{min,max}_bounds``.
* We now have six spatial types, ``Point``, ``LineString`` and ``Polygon``

spyne/error.py

#
# spyne - Copyright (C) Spyne contributors.
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
#


"""The ``spyne.error`` module contains various common exceptions that the user
code can throw.
"""


from spyne.model.fault import Fault
from spyne.const import MAX_STRING_FIELD_LENGTH


class InvalidCredentialsError(Fault):
    """Raised when requested resource is not found."""

    def __init__(self, fault_object,
             fault_string="You do not have permission to access this resource"):
        Fault.__init__(self, 'Client.InvalidCredentialsError', fault_string)


class RequestTooLongError(Fault):
    """Raised when request is too long."""

    def __init__(self, faultstring="Request too long"):
        Fault.__init__(self, 'Client.RequestTooLong', faultstring)


class RequestNotAllowed(Fault):
    """Raised when request is incomplete."""

    def __init__(self, faultstring=""):
        Fault.__init__(self, 'Client.RequestNotAllowed', faultstring)


class ArgumentError(Fault):
    """Raised when there is a general problem with input data."""

    def __init__(self, faultstring=""):
        Fault.__init__(self, 'Client.ArgumentError', faultstring)


class InvalidInputError(Fault):
    """Raised when there is a general problem with input data."""

    def __init__(self, faultstring="", data=""):
        Fault.__init__(self, 'Client.InvalidInput', repr((faultstring, data)))

InvalidRequestError = InvalidInputError

class ValidationError(Fault):
    """Raised when the input stream does not adhere to type constraints."""

    def __init__(self, obj, custom_msg='The value %r could not be validated.'):
        s = repr(obj)

        if len(s) > MAX_STRING_FIELD_LENGTH:
            s = s[:MAX_STRING_FIELD_LENGTH] + "(...)"

        Fault.__init__(self, 'Client.ValidationError', custom_msg % s)


class InternalError(Fault):
    """Raised to communicate server-side errors."""
    def __init__(self, error):
        Fault.__init__(self, 'Server', "InternalError: An unknown error has occured.")


class ResourceNotFoundError(Fault):
    """Raised when requested resource is not found."""

    def __init__(self, fault_object,
                 fault_string="Requested resource %r not found"):
        Fault.__init__(self, 'Client.ResourceNotFound', fault_string %
                       fault_object)


class ResourceAlreadyExistsError(Fault):
    """Raised when requested resource already exists on server side."""

    def __init__(self, fault_object,
                 fault_string="Resource %r already exists"):
        Fault.__init__(self, 'Client.ResourceAlreadyExists', fault_string %
                       fault_object)


spyne/__init__.py

#
# spyne - Copyright (C) Spyne contributors.
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
#

__version__ = '2.11.0.dev1'

from pytz import utc as LOCAL_TZ

from spyne._base import BODY_STYLE_WRAPPED
from spyne._base import BODY_STYLE_BARE
from spyne._base import AuxMethodContext
from spyne._base import TransportContext
from spyne._base import EventContext
from spyne._base import MethodContext
from spyne._base import MethodDescriptor
from spyne._base import EventManager

from spyne.decorator import rpc
from spyne.decorator import srpc
from spyne.service import ServiceBase
from spyne.application import Application

import sys

if not hasattr(sys, "version_info") or sys.version_info < (2, 6):
    raise RuntimeError("Spyne requires Python 2.6 or later.")

del sys


spyne/model/_base.py

#
# spyne - Copyright (C) Spyne contributors.
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
#

import re

import spyne.const.xml_ns

from decimal import Decimal

from spyne.util import Break

"""This module contains the ModelBase class and other building blocks for
defining models.
"""


def nillable_string(func):
    """Decorator that retuns None if input is None."""

    def wrapper(cls, string, *args, **kwargs):
        if string is None:
            return None
        else:
            return func(cls, string, *args, **kwargs)
    return wrapper


def nillable_iterable(func):
    """Decorator that retuns [] if input is None."""

    def wrapper(prot, cls, string):
        if string is None:
            return []
        else:
            return func(prot, cls, string)
    return wrapper


class AttributesMeta(type(object)):
    """I hate quirks. This is a 10-minute attempt to get rid of a one-letter
    quirk."""

    def __new__(cls, cls_name, cls_bases, cls_dict):
        # Mapper args should not be inherited.
        if not 'sqla_mapper_args' in cls_dict:
            cls_dict['sqla_mapper_args'] = None

        return type(object).__new__(cls, cls_name, cls_bases, cls_dict)

    def __init__(self, cls_name, cls_bases, cls_dict):
        nullable = cls_dict.get('nullable', None)
        nillable = cls_dict.get('nillable', None)

        assert nullable is None or nillable is None or nullable == nillable

        self.__nullable = nullable or nillable or True

        type(object).__init__(self, cls_name, cls_bases, cls_dict)

    def get_nullable(self):
        return self.__nullable
    def set_nullable(self, what):
        self.__nullable = what
    nullable = property(get_nullable, set_nullable)

    def get_nillable(self):
        return self.__nullable
    def set_nillable(self, what):
        self.__nullable = what
    nillable = property(get_nillable, set_nillable)


class ModelBase(object):
    """The base class for type markers. It defines the model interface for the
    interface generators to use and also manages class customizations that are
    mainly used for defining constraints on input values.
    """

    __orig__ = None
    """This holds the original class the class .customize()d from. """

    __extends__ = None
    """This holds the original class the class inherited or .customize()d from.
    This is different from __orig__ because it's only set when
    ``cls.is_default(cls) == False``"""

    __namespace__ = None
    """The public namespace of this class. Use ``get_namespace()`` instead of
    accessing it directly."""

    __type_name__ = None
    """The public type name of the class. Use ``get_type_name()`` instead of
    accessing it directly."""

    # These are not the xml schema defaults. The xml schema defaults are
    # considered in XmlSchema's add() method. the defaults here are to reflect
    # what people seem to want most.
    #
    # Please note that min_occurs and max_occurs must be validated in the
    # ComplexModelBase deserializer.
    class Attributes(object):
        """The class that holds the constraints for the given type."""

        __metaclass__ = AttributesMeta

        _wrapper = False
        # when skip_wrappers=True is passed to a protocol, these objects
        # are skipped. just for internal use.

        default = None
        """The default value if the input is None"""

        nillable = True
        """Set this to false to reject null values. Synonyms with
        ``nullable``."""

        min_occurs = 0
        """Set this to 1 to make this object mandatory. Can be set to any
        positive integer. Note that an object can still be null or empty, even
        if it's there."""

        max_occurs = 1
        """Can be set to any strictly positive integer. Values greater than 1
        will imply an iterable of objects as native python type. Can be set to
        ``decimal.Decimal("inf")`` for arbitrary number of arguments."""

        schema_tag = '{%s}element' % spyne.const.xml_ns.xsd
        """The tag used to add a primitives as child to a complex type in the
        xml schema."""

        translations = None
        """A dict that contains locale codes as keys and translations of field
        names to that language as values.
        """

        sub_ns = None
        """An Xml-specific attribute that specifies which namespace should be
        used for field names in classes.
        """

        sub_name = None
        """This specifies which string should be used as field name when this
        type is seriazed under a ComplexModel.
        """

        sqla_column_args = None
        """A dict that will be passed to SQLAlchemy's ``Column`` constructor as
        ``**kwargs``.
        """

        exc_mapper = False
        """If true, this field will be excluded from the table mapper of the
        parent class.
        """

        exc_table = False
        """If true, this field will be excluded from the table of the parent
        class.
        """

        exc_interface = False
        """If true, this field will be excluded from the interface document."""

        logged = True
        """If false, this object will be ignored in ``log_repr``, mostly used
        for logging purposes."""

        unique = None
        """If True, this object will be set as unique in the database schema
        with default indexing options. If the value is a string, it will be
        used as the indexing method to create the unique index. See sqlalchemy
        documentation on how to create multi-column unique constraints.
        """

        db_type = None
        """When not None, it overrides Spyne's own mapping from Spyne types to
        SQLAlchemy types. It's a standard SQLAlchemy type marker, e.g.
        ``sqlalchemy.Integer``.
        """

        table_name = None
        """Database table name."""

        xml_choice_group = None
        """When not None, shares the same <choice> tag with fields with the same
        xml_choice_group value.
        """

        index = None
        """Can be ``True``, a string, or a tuple of two strings.

        * If True, this object will be set as indexed in the database schema
          with default options.

        * If the value is a string, the value will denote the indexing method
          used by the database. See: http://www.postgresql.org/docs/9.2/static/indexes-types.html

        * If the vale is a tuple of two strings, the first value will denote the
          index name and the second value will denote the indexing method as
          above.
        """

        read_only= False
        """If True, the attribute won't be initialized from outside values."""

    class Annotations(object):
        """The class that holds the annotations for the given type."""

        __use_parent_doc__ = False
        """If set to True Annotations will use __doc__ from parent,
        This is a convenience option"""

        doc = ""
        """The public documentation for the given type."""

        appinfo = None
        """Any object that carries app-specific info."""

    class Empty(object):
        pass

    _force_own_namespace = None

    @staticmethod
    def is_default(cls):
        return True

    @classmethod
    def get_namespace_prefix(cls, interface):
        """Returns the namespace prefix for the given interface. The
        get_namespace_prefix of the interface class generates a prefix if none
        is defined.
        """

        ns = cls.get_namespace()

        retval = interface.get_namespace_prefix(ns)

        return retval

    @classmethod
    def get_namespace(cls):
        """Returns the namespace of the class. Defaults to the python module
        name."""

        return cls.__namespace__

    @staticmethod
    def resolve_namespace(cls, default_ns, tags=None):
        """This call finalizes the namespace assignment. The default namespace
        is not available until the application calls populate_interface method
        of the interface generator.
        """

        if cls.__namespace__ is spyne.const.xml_ns.DEFAULT_NS:
            cls.__namespace__ = default_ns

        if (cls.__namespace__ in spyne.const.xml_ns.const_prefmap and
                                                       not cls.is_default(cls)):
            cls.__namespace__ = default_ns

        if cls.__namespace__ is None:
            ret = []
            for f in cls.__module__.split('.'):
                if f.startswith('_'):
                    break
                else:
                    ret.append(f)

            cls.__namespace__ = '.'.join(ret)

        if cls.__namespace__ is None or len(cls.__namespace__) == 0:
            cls.__namespace__ = default_ns

        if cls.__namespace__ is None or len(cls.__namespace__) == 0:
            raise ValueError("You need to explicitly set %r.__namespace__" % cls)

    @classmethod
    def get_type_name(cls):
        """Returns the class name unless the __type_name__ attribute is defined.
        """

        retval = cls.__type_name__
        if retval is None:
            retval = cls.__name__

        return retval

    @classmethod
    def get_type_name_ns(cls, interface):
        """Returns the type name with a namespace prefix, separated by a column.
        """

        if cls.get_namespace() != None:
            return "%s:%s" % (cls.get_namespace_prefix(interface), cls.get_type_name())

    @classmethod
    @nillable_string
    def to_string(cls, value):
        """Returns str(value). This should be overridden if this is not enough.
        """

        return str(value)

    @classmethod
    def customize(cls, **kwargs):
        """Duplicates cls and overwrites the values in ``cls.Attributes`` with
        ``**kwargs`` and returns the new class."""

        cls_name, cls_bases, cls_dict = cls._s_customize(cls, **kwargs)

        return type(cls_name, cls_bases, cls_dict)

    @staticmethod
    def _s_customize(cls, **kwargs):
        """This function duplicates and customizes the class it belongs to. The
        original class remains unchanged.

        Not meant to be overridden.
        """

        cls_dict = {}
        if getattr(cls, '__orig__', None) is None:
            cls_dict['__orig__'] = cls

        class Attributes(cls.Attributes):
            pass

        if cls.Attributes.translations is None:
            Attributes.translations = {}
        if cls.Attributes.sqla_column_args is None:
            Attributes.sqla_column_args = (), {}

        cls_dict['Attributes'] = Attributes

        # as nillable is a property, it gets reset everytime a new class is
        # defined. So we need to reinitialize it explicitly.
        Attributes.nillable = cls.Attributes.nillable

        class Annotations(cls.Annotations):
            pass
        cls_dict['Annotations'] = Annotations

        for k, v in kwargs.items():
            if k.startswith('_'):
                continue

            elif k in ("doc", "appinfo"):
                setattr(Annotations, k, v)

            elif k in ('primary_key','pk'):
                Attributes.sqla_column_args[-1]['primary_key'] = v

            elif k in ('foreign_key','fk'):
                from sqlalchemy.schema import ForeignKey
                t, d = Attributes.sqla_column_args
                fkt = (ForeignKey(v),)
                Attributes.sqla_column_args = (t + fkt, d)

            elif k in ('autoincrement', 'onupdate', 'server_default'):
                Attributes.sqla_column_args[-1][k] = v

            elif k == 'max_occurs' and v == 'unbounded':
                setattr(Attributes, k, Decimal('inf'))

            else:
                setattr(Attributes, k, v)

        return (cls.__name__, (cls,), cls_dict)

    @staticmethod
    def validate_string(cls, value):
        """Override this method to do your own input validation on the input
        string. This is called before converting the incoming string to the
        native python value."""

        return (cls.Attributes.nillable or value is not None)

    @staticmethod
    def validate_native(cls, value):
        """Override this method to do your own input validation on the native
        value. This is called after converting the incoming string to the
        native python value."""

        return (cls.Attributes.nullable or value is not None)


class Null(ModelBase):
    pass


class SimpleModel(ModelBase):
    """The base class for primitives."""

    __namespace__ = "http://www.w3.org/2001/XMLSchema"

    class Attributes(ModelBase.Attributes):
        """The class that holds the constraints for the given type."""

        class __metaclass__(AttributesMeta):
            def __init__(self, cls_name, cls_bases, cls_dict):
                AttributesMeta.__init__(self, cls_name, cls_bases, cls_dict)
                if getattr(self, '_pattern', None) is None:
                    self._pattern = None
            def get_pattern(self):
                return self._pattern
            def set_pattern(self, pattern):
                self._pattern = pattern
                if pattern is not None:
                    self.pattern_re = re.compile(pattern)
            pattern = property(get_pattern, set_pattern)

        values = set()
        """The set of possible values for this type."""

        pattern_re = None

    def __new__(cls, **kwargs):
        """Overriden so that any attempt to instantiate a primitive will return
        a customized class instead of an instance.

        See spyne.model.base.ModelBase for more information.
        """

        retval = cls.customize(**kwargs)

        if not retval.is_default(retval):
            retval.__extends__ = cls
            retval.__type_name__ = kwargs.get("type_name", ModelBase.Empty)

        return retval

    @staticmethod
    def is_default(cls):
        return (cls.Attributes.values == SimpleModel.Attributes.values)

    @staticmethod
    def validate_string(cls, value):
        return (     ModelBase.validate_string(cls, value)
                and (len(cls.Attributes.values) == 0 or
                                                value in cls.Attributes.values)
            )


class PushBase(object):
    def __init__(self, callback, errback=None):
        self._cb = callback
        self._eb = errback

        self.length = 0
        self.ctx = None
        self.app = None
        self.response = None
        self.gen = None
        self._cb_finish = None
        self._eb_finish = None

    def _init(self, ctx, response, gen, _cb_finish, _eb_finish):
        self.length = 0

        self.ctx = ctx
        self.app = ctx.app

        self.response = response
        self.gen = gen

        self._cb_finish = _cb_finish
        self._eb_finish = _eb_finish

    def init(self, ctx, response, gen, _cb_finish, _eb_finish):
        self._init(ctx, response, gen, _cb_finish, _eb_finish)
        return self._cb(self)

    def __len__(self):
        return self.length

    def append(self, inst):
        self.gen.send(inst)
        self.length += 1

    def close(self):
        try:
            self.gen.throw(Break())
        except StopIteration:
            pass

        self._cb_finish()
